### PR TITLE
Create rule S6658: Special methods should have an expected return type

### PR DESCRIPTION
**Pull request by maksim-grebeniuk-sonarsource**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S6658/python/rule.adoc [hunk 1]
```diff
@@ -0,0 +1,83 @@
+This rule raises an issue when a special method is defined with an unexpected number of parameters.
```
Comment (guillaume-dequenne-sonarsource) [inline rules/S6658/python/rule.adoc:1]: This should be updated:
`This rule raises an issue when a special method returns an object of an unexpected type.`

File: rules/S6658/python/rule.adoc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+== Why is this an issue?
+
+Python allows developers to customize how code is interpreted by defining special methods (also called magic methods). For example, it is possible to define the objects own truthiness or falsiness by iverriding ``__bool__`` method. It is invoked when the built-in ``bool()`` function is called on an object. The ``bool()`` function returns ``True`` or ``False`` based on the truth value of the object.
+
+Each special method expects a specific return type. The Python interpreter will call these methods. Calls to a special method will throw a ``++TypeError++`` if it there is returning value with incorrect type.
```
Comment (guillaume-dequenne-sonarsource) [inline rules/S6658/python/rule.adoc:7]: I'd probably complete:
`The Python interpreter will call these methods when performing the operation they're associated with.`
And I'd rephrase:
`Calls to a special method will throw a ``++TypeError++`` if its return type is incorrect.`

File: rules/S6658/python/rule.adoc [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+== How to fix it
+
+Make sure to return value with correct type defined in the Python documentation for each special methods.
```
Comment (guillaume-dequenne-sonarsource) [inline rules/S6658/python/rule.adoc:11]: I'd rephrase:
`Make sure to return a value of the same type as defined in the Python documentation for each special method.`

File: rules/S6658/python/rule.adoc [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+=== Code examples
+
+==== Noncompliant code example
+
+[source,python,diff-id=1,diff-type=noncompliant]
+----
+class MyClass:
+    def __bool__(self):
+        return 0 # Noncompliant: incorrect return type. Boolean value expected.
```
Comment (guillaume-dequenne-sonarsource) [inline rules/S6658/python/rule.adoc:21]: I'd say: `# Noncompliant: returned value should be of type bool.` to consistently used the exact type name.

File: rules/S6658/python/rule.adoc [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+obj1 = MyClass()
+print(bool(obj1))
+----
+
+== Resources
+
+=== Documentation
+
+* https://docs.python.org/3/reference/datamodel.html#special-method-names[Special method names] - Python special methods 
+* https://docs.python.org/3/library/copy.html[Copy module] - Documentation of ``++__copy__++`` and ``++__deepcopy__++``
```
Comment (guillaume-dequenne-sonarsource) [inline rules/S6658/python/rule.adoc:44]: We should make the links consistent with the new format (Python Documentation - ...)

File: rules/S6658/python/rule.adoc [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+ifdef::env-github,rspecator-view[]
+
+'''
+== Implementation Specification
+(visible only on this page)
+
+=== Message
+
+* Make sure this method returns a value of type XXX.
```
Comment (guillaume-dequenne-sonarsource) [inline rules/S6658/python/rule.adoc:55]: I guess we can even say `Return a value of type XXX here.` to be even more direct in the issue message.

File: rules/S6658/python/rule.adoc [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+==== if there is no return statement at all
+    Primary: method name
+
+
+'''
+== Comments And Links
+(visible only on this page)
+
+=== Methods to be checked:
```
Comment (guillaume-dequenne-sonarsource) [inline rules/S6658/python/rule.adoc:72]: Probably we can add these methods to the `Why is this an issue part`, saying that an issue will be raised when one of the following method doesn't return the indicated type. 
That would probably make easier for people to understand when they can expect an issue (since all special methods are not listed here).

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Looking good, left a few comments.

File: rules/S6658/python/rule.adoc [hunk 8]
```diff
@@ -1,14 +1,26 @@
-This rule raises an issue when a special method is defined with an unexpected number of parameters.
+This rule raises an issue when a special method returns an object of an unexpected type.
 
 == Why is this an issue?
 
 Python allows developers to customize how code is interpreted by defining special methods (also called magic methods). For example, it is possible to define the objects own truthiness or falsiness by iverriding ``__bool__`` method. It is invoked when the built-in ``bool()`` function is called on an object. The ``bool()`` function returns ``True`` or ``False`` based on the truth value of the object.
 
-Each special method expects a specific return type. The Python interpreter will call these methods. Calls to a special method will throw a ``++TypeError++`` if it there is returning value with incorrect type.
+The Python interpreter will call these methods when performing the operation they're associated with. Each special method expects a specific return type. The Python interpreter will call these methods. Calls to a special method will throw a ``++TypeError++`` if its return type is incorrect.
```
Comment (guillaume-dequenne-sonarsource) [inline rules/S6658/python/rule.adoc:7]: It probably doesn't make sense to repeat `The Python interpreter will call these methods.` in the second part of the explanation.

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Left a few comments.

File: rules/S6658/python/rule.adoc [hunk 9]
```diff
@@ -0,0 +1,82 @@
+This rule raises an issue when a special method returns an object of an unexpected type.
+
+== Why is this an issue?
+
+Python allows developers to customize how code is interpreted by defining special methods (also called magic methods). For example, it is possible to define the objects own truthiness or falsiness by iverriding ``++__bool__++`` method. It is invoked when the built-in ``++bool()++`` function is called on an object. The ``++bool()++`` function returns ``++True++`` or ``++False++`` based on the truth value of the object.
```
Comment (anton-haubner-sonarsource) [inline rules/S6658/python/rule.adoc:5]: Minor issue: Typo: "iverriding" instead of "overriding".

File: rules/S6658/python/rule.adoc [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+obj1 = MyClass()
+print(bool(obj1))
+----
+
+== Resources
+
+=== Documentation
+
+* Python Documentation - https://docs.python.org/3/reference/datamodel.html#special-method-names[Special method names]
+
```
Comment (anton-haubner-sonarsource) [inline rules/S6658/python/rule.adoc:56]: Since the spec also mentions `__getnewargs__` and `__getnewargs_ex__`, maybe we should also point to their documentation at https://docs.python.org/3/library/pickle.html#pickling-class-instances
